### ISSUE TITLE
Public API of the framework

### ISSUE THREAD

**Issue by wtimme** [holyturt/SwiftOverpassWrapper#3]
Hey @holyturt,

Thanks for your effort, and thanks for accepting the pull requests!

My usecase for the framework is querying the Overpass server for nodes in a bounding box, basically for performing the bare requests. I store nodes in memory and keep track of them in memory, merging new ones with existing ones, avoiding duplicates, etc by myself. Therefore, I'm not dependend on the methods for loading related entities, such as `loadRelatedWays()`.

**How do you use this framework?** I'm asking because I'd like to exchange ideas and opinions. Do you have any experience that you can share? Thanks in advance for your response!

**Comment by holyturt** [comment]
Hi @wtimme,
Thank you for your message and sorry for the delay. The notification seems not working.
Your pull requests were really helpful!

My last project needed **relations** of city boundary and **nodes** which are related to the **relations** to overlay it on a map. Also, I didn't want to write raw overpass queries on source files as multiple lines of String and I believed it will be more readable. Therefore, I made this framework. My goal was to write query easily in Swift. Miscellaneous methods, such as `loadRelatedWays()`, were made by such aspect. 
I think users don't want to use their time to learn Overpass QL or XML just to get map data for iOS development.

Honestly, this framework has lack of specific visions and usecases. And this is why this repository is currently lazy.
So, I'm very happy to hear your ideas and opinions! I'd like to improve the framework with it! Thanks.

**Comment by wtimme** [comment]
Thanks for the feedback and the insights, @holyturt!

I'm currently working on an app that displays certain **nodes** on a map, loading the information lazily as the users pans on the map. Furthermore, the app allows for editing the tags of the POIs on the map. The purpose is to give users the possibility to add small bits of information that is missing from those POIs, e. g. the numbers of parking spots available at a parking space, in a similar manner to the [quests](https://wiki.openstreetmap.org/wiki/StreetComplete/Quests) of the Android app [StreetComplete](https://wiki.openstreetmap.org/wiki/StreetComplete).

I'm glad that you're open for feedback. Here's the first thing that pops into my mind. Let's get the conversation rolling! :)

# Testing
I'm a big fan of unit testing, and try to embrace TDD. When new to a project, adding tests help me getting to know the API and understanding the underlying logic. Furthermore, it helps me being more confident with refactorings, as the tests will have my back if I screw up. 😉
I've already started adding tests for the XML parsers, but noticed that in order to create the node, entity or way, I needed to always provide the `OverpassResponse` that the XML came with. This doesn't feel quite right. Have a look at this snippet from one of the test cases. `xmlElement` is a `AEXMLElement` that I read from a local file.

```
let dataResponse = DataResponse<String>(request: nil,
                                        response: nil,
                                        data: Data(),
                                        result: Result<String>.success(""))

let response = OverpassResponse(response: dataResponse,
                                requestQuery: "")

// Actually use the XML parser
guard let node = OverpassNode(xmlElement: xmlElement, response: response) else {
    XCTFail("The XML should properly initialize the model.")
    return
}
```
So just for testing the XML parsing, I need to create an `OverpassResponse` that I don't actually need for the XML parsing. But when wanting to lookup releations, the models *need* to have a reference to the `response`. So what could be a solution?

## Dedicated object for relations lookup
Instead of having the reference to the `response` that an item was retrieved with, I suggest to use a dedicated object that can be queried for this information. Internally, it would keep track of the "world" that it knows about, and could lazily lookup data when needed.
This serves three purposes:
- Keeping the models as stupid possible, having only actual data attached to them. This makes unit testing easier and reduces the complexity of the models. Additionally, it improves readability, which is always good to have, as there's no logic in those objects.
- Letting the app "learn" about the world: With each query, new relations might be discovered, and "old" references to `response` instances don't contain the whole picture.
- Preparing/allowing for the addition of optional caching: This is something that I'm currently implementing in my own app, but would be nice to have shared with everyone.

# Other ideas for the framework
Beside improving the testability and the coverage, I have a couple of feature proposals/ideas.

## Caching

Maybe using a database such as SQLite. Could be useful for offline applications. If we didn't want to include a cache in _this_ project, we could still make sure that our objects are easy to cache and that a cache can optionally be added to the data management object.

## Elaborate "rules"
This is something that _StreetComplete_ has and what is really nice for reading/editing queries. Have for example a look at [the tag filter for the `AddBabyChangingTable` quest](https://github.com/westnordost/StreetComplete/blob/205cd4ca8461a782ffcba9d875beefd5c6c53348/app/src/main/java/de/westnordost/streetcomplete/quests/baby_changing_table/AddBabyChangingTable.java#L23):
```
(type:node or type:way)
and
diaper!=*
and
(
    amenity = toilets
    or
    (
        (
            amenity~="restaurant|cafe|fuel|fast_food"
            or
            shop~="mall|department_store"
        )
        and
        name=*
        and
        toilets=yes
    )
)
```
As an added bonus, the _StreetComplete_ has a [test case for their parser](https://github.com/westnordost/StreetComplete/blob/00b5d17af222a73b84c9b7eb733a34e42e61d1d0/app/src/test/java/de/westnordost/streetcomplete/data/osm/tql/FiltersParserTest.java), allowing for TDD. 🤓
A proper parser allows for more complex rules than we can currently do in code. Furthermore, the queries could be built using an external tool (such as _overpass turbo_) and can be exchanged.

I guess for starters, this is enough information. 😄 Thanks for taking the time! I'm very much looking forward to your feedback.

**Comment by holyturt** [comment]
First of all, thank you for your ideas and proposals, @wtimme! It is really helpful and inspires me. Your project and the StreetComplete as well. Thanks for sharing it!

## Testing
I'm glad to hear that you are a unit test enthusiast. I'm the type of guy who only have knowledge but few experience about unit testing. Therefore, I will merge your tests with pleasure!

## Dedicated object for relations lookup
As you pointed out, `OverpassNode` and the others shouldn't involve the `OverpassResponse`. It's embarrassing. 😅
The idea is really great. As you mentioned, it will make those models so simple and readable! Furthermore, it would provides flexibility for changes near future. I'm happy that `OverpassNode` no longer have unrelated lookup methods such as `loadRelatedRelations()`!
So do you have any ideas you can share for implementing the dedicated object from your works or thoughts?

## Caching
Caching data seems not necessary for every apps. But it is an important feature for many apps which this project relate to. So, in my opinion, it should be optional.
There are several ways to cache data in iOS development such as SQLite, Realm. Do you have any ideas how to help user to cache easily with this project?

## Elaborate rules
The StreetComplete's query looks simple and functional. It would be nice to have such "rules" in this project! Because this project supports few filters now, e.g. bbox, tag, and it may have more (complex) filters. It will work tremendously in future! Implementing complex parser seems to cost much time but it is worth to do.

## Overpass XML vs. Overpass QL
I don't know this topic is important but what do you think about using Overpass XML for the API. When I started this project, I picked it without thinking so much. However, if using Overpass QL is more convenient in future, I could switch to it. Because the project seems still simple enough to do it.

Lastly, could you be a collaborator? It would be nice if you join the project! Please feel free to ask anything! 😉 Thanks.

**Comment by wtimme** [comment]
Thanks for your thorough response, @holyturt!

## Dedicated object for relations lookup
> So do you have any ideas you can share for implementing the dedicated object from your works or thoughts?

Yes, I do. First of all, I'd create a central place for storing the results of all the responses so far, and already started working on a protocol. Take a look:
```swift
public protocol OverpassEntityManaging {
    /// Provides all available entities from storage.
    var entities: [OverpassEntity] { get }
    
    /// Adds a list of entities to the storage.
    /// The method will filter and list and only add those nodes that are either completely new or updated.
    ///
    /// - Parameter entities: The entities to process.
    func add(_ entities: [OverpassEntity])
}
```
As you can see, the object takes elements of type `OverpassEntity` and processes them. In my current implementation, `add(_:)` already handles multiple instances of the same entity being added, and maintains a list of only the newer ones (= those with a greater version).
I plan on having yet another object that contains a `OverpassEntityManaging` object for each of the elements (way, node and relation), but haven't started implementing it yet. That object can then have methods for looking up related nodes, ways and relations. I'll come back to you once I have some results to show.

### Caching
> Caching data seems not necessary for every apps. But it is an important feature for many apps which this project relate to. So, in my opinion, it should be optional.

Yes, I agree. That's why it would be nice to work with protocols, as it allows other developers to _optionally_ add their own cache. I haven't worked on this yet, though, because I feel like the first point (having a dedicated object for relations lookup) needs to be addressed before we can move on to caching.

### Elaborate rules
This is something for the future, and could even reside in its own little project, as _SwiftOverpassWrapper_ doesn't necessarily depend on it. Before adding this new feature, I'd make sure that the rest of the code base is in a way that can be easily extended.

### Overpass XML vs. Overpass QL
I haven't had a closer look on the query language vs. the XML, but if we go with the rules it may save us a conversion step if we also decide to switch to Overpass QL.

> Lastly, could you be a collaborator? It would be nice if you join the project!

I'd happily do so. Is there anything from my end that you need, or is it something that you can set up in GitHub?

On an ending note: How about we move the different topics to their dedicated issues? That way, it's easier for other contributors to have a look at the plans/ideas for _SwiftOverpassWrapper_.

**Comment by holyturt** [comment]
Moved these topics to another. And I sent a invitation. Please check it. 😉
